File: python/src/schema_audit/_per_item.py
"""Per-item validation logic — shared across format orchestrators.

Cycle 9 has only one orchestrator (JSON-LD), but the per-item engine is
extracted now so cycle 10's Microdata + RDFa orchestrators can plug in
cleanly — exactly as the JS side did in cycle 6.

Pure: no I/O, no shared mutable state. The registry + curated rules are
loaded once at module import.

Mirrors ``core/src/validators/per-item.ts`` 1:1.
"""

from __future__ import annotations

from typing import NamedTuple

from .curated_rules import get_curated_rules
from .registry import load_registry
from .rules.validate_context import validate_context
from .rules.validate_property_existence import validate_property_existence
from .rules.validate_property_value_type import validate_property_value_type
from .rules.validate_recommended import validate_recommended
from .rules.validate_required import validate_required
from .rules.validate_type import validate_type
from .rules.validate_url import validate_url
from .types import Issue

_REGISTRY = load_registry()

# JSON-LD reserved keys (and the Microdata diagnostic field reserved for
# cycle 10) — never validated as schema.org properties.
_RESERVED_KEYS = frozenset(
    {
        "@context",
        "@type",
        "@id",
        "@graph",
        "@language",
        "@reverse",
        "@vocab",
        "@base",
        "@itemtypeRaw",
    }
)

_STRING_NON_URL_PRIMITIVES = frozenset({"Text", "Date", "DateTime", "Time"})


def _should_validate_string_as_url(value_types: list[str]) -> bool:
    if "URL" not in value_types:
        return False
    return not any(t in _STRING_NON_URL_PRIMITIVES for t in value_types)


class PerItemResult(NamedTuple):
    issues: list[Issue]
    resolved_type: str | None


def validate_item(obj: dict[str, object]) -> PerItemResult:
    """Validate a single JSON-LD-shaped item.

    The caller is responsible for parse-time handling and asserting the
    input is a non-null, non-list dict.
    """
    issues: list[Issue] = []

    # 1. @context
    issues.extend(validate_context(obj.get("@context")))

    # 2. @type
    type_value = obj.get("@type")
    issues.extend(validate_type(type_value, _REGISTRY))

    resolved_type: str | None = None
    if isinstance(type_value, str) and type_value in _REGISTRY["types"]:
        resolved_type = type_value
        type_def = _REGISTRY["types"][type_value]

        # 3. Properties
        for key, val in obj.items():
            if key in _RESERVED_KEYS:
                continue

            existence = validate_property_existence(type_value, key, val, _REGISTRY)
            if existence:
                issues.extend(existence)
                continue

            prop_def = type_def["allProperties"][key]
            value_types = prop_def["valueTypes"]
            issues.extend(
                validate_property_value_type(type_value, key, val, value_types)
            )

            if isinstance(val, str) and _should_validate_string_as_url(value_types):
                issues.extend(validate_url(type_value, key, val))

        # 4. Layer 2 — Google Rich Results
        l2 = get_curated_rules(type_value)
        if l2 is not None:
            issues.extend(
                validate_required(obj, type_value, l2["required"], l2["requiredOneOf"])
            )
            issues.extend(validate_recommended(obj, type_value, l2["recommended"]))

    return PerItemResult(issues=issues, resolved_type=resolved_type)

File: core/src/validators/per-item.ts
// core/src/validators/per-item.ts
//
// Per-item validation logic — shared by the JSON-LD and Microdata
// orchestrators (and later, RDFa). Takes a JSON-LD-shaped object
// (parsed and known to be an object); runs Layer 1 (schema.org
// structural) + Layer 2 (Google Rich Results); returns Issue[].
//
// This file is the heart of the validator's per-item rules. The
// surrounding orchestrators handle format-specific concerns
// (parsing JSON, extracting Microdata) and then funnel each item
// through here.
//
// Pure: no I/O, no shared mutable state. The registry + curated
// rules are loaded once at module init.

import type { Issue } from "../types.js";
import { loadRegistry } from "../registry.js";
import { getCuratedRules } from "../curated-rules.js";
import { validateContext } from "../rules/validate-context.js";
import { validateType } from "../rules/validate-type.js";
import { validatePropertyExistence } from "../rules/validate-property-existence.js";
import { validatePropertyValueType } from "../rules/validate-property-value-type.js";
import { validateUrl } from "../rules/validate-url.js";
import { validateRequired } from "../rules/validate-required.js";
import { validateRecommended } from "../rules/validate-recommended.js";

const registry = loadRegistry();

// JSON-LD reserved keys (and Microdata's `@itemtypeRaw` diagnostic
// field, which never represents a real property).
const RESERVED_KEYS = new Set([
  "@context",
  "@type",
  "@id",
  "@graph",
  "@language",
  "@reverse",
  "@vocab",
  "@base",
  "@itemtypeRaw",
]);

const STRING_NON_URL_PRIMITIVES = new Set([
  "Text",
  "Date",
  "DateTime",
  "Time",
]);

function shouldValidateStringAsUrl(valueTypes: string[]): boolean {
  if (!valueTypes.includes("URL")) return false;
  for (const t of valueTypes) {
    if (STRING_NON_URL_PRIMITIVES.has(t)) return false;
  }
  return true;
}

export interface PerItemResult {
  issues: Issue[];
  /** The @type value if it resolved to a registered type, else null. */
  resolvedType: string | null;
}

/**
 * Validates a single JSON-LD-shaped item.
 *
 * @param o The parsed item object (already known to be a non-null,
 *          non-array object). Caller is responsible for parse-time
 *          handling and shape validation.
 * @returns issues found and the resolved type (or null when
 *          @type is missing/unknown).
 */
export function validateItem(o: Record<string, unknown>): PerItemResult {
  const issues: Issue[] = [];

  // Context
  issues.push(...validateContext(o["@context"]));

  // Type
  const typeValue = o["@type"];
  issues.push(...validateType(typeValue, registry));

  let resolvedType: string | null = null;
  if (
    typeof typeValue === "string" &&
    registry.types[typeValue] !== undefined
  ) {
    resolvedType = typeValue;
    const typeDef = registry.types[typeValue];

    // Properties
    for (const [key, val] of Object.entries(o)) {
      if (RESERVED_KEYS.has(key)) continue;

      const existence = validatePropertyExistence(
        typeValue,
        key,
        val,
        registry,
      );
      if (existence.length > 0) {
        issues.push(...existence);
        continue;
      }

      const propDef = typeDef.allProperties[key]!;
      issues.push(
        ...validatePropertyValueType(typeValue, key, val, propDef.valueTypes),
      );

      if (
        typeof val === "string" &&
        shouldValidateStringAsUrl(propDef.valueTypes)
      ) {
        issues.push(...validateUrl(typeValue, key, val));
      }
    }

    // Layer 2 — Google Rich Results
    const l2 = getCuratedRules(typeValue);
    if (l2 !== undefined) {
      issues.push(
        ...validateRequired(o, typeValue, l2.required, l2.requiredOneOf),
      );
      issues.push(...validateRecommended(o, typeValue, l2.recommended));
    }
  }

  return { issues, resolvedType };
}
